src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java

```diff
@@ -1080,6 +1081,32 @@ private String checkMaxOrphanedHeadersExceeded(HeaderFrame hf) {
         return null;
     }
 
+    // This method is called when a DataFrame that was added
+    // to a Stream::inputQ is later dropped from the queue
+    // without being consumed.
+    //
+    // Before adding a frame to the queue, the Stream calls
+    // connection.windowUpdater.canBufferUnprocessedBytes(), which
+    // increases the count of unprocessed bytes in the connection.
+    // After consuming the frame, it calls connection.windowUpdater::processed,
+    // which decrements the count of unprocessed bytes, and possibly
+    // sends a window update to the peer.
+    //
+    // This method is called when connection.windowUpdater::processed
+    // will not be called, which can happen when consuming the frame
+    // fails, or when an empty DataFrame terminates the stream,
+    // or when the stream is cancelled while data is still
+    // sitting in its inputQ. In the later case, it is called for
+    // each frame that is dropped from the queue.
+    final void releaseUnconsumed(DataFrame df) {
+        windowUpdater.released(df.payloadLength());
+        dropDataFrame(df);
+    }
+
+    // This method is called when a DataFrame is dropped before/without
+    // having been added to any Stream::inputQ. In that case, the number
+    // of unprocessed bytes hasn't been incremented by the stream, and
+    // does not need to be decremented.
```

Comment:
Maybe I should say "is called directly" ? The idea is that if the data is dropped before being added to the inputQ, then `dropDataFrame` is called directly. If the data is dropped after having been added to the inputQ, then `releaseUnconsumed` is called.

Reply:
> Maybe I should say "is called directly" ? 

I think that would still be unclear since "directly" is contextual and depends on who is calling it. So we would then have to clarify what a direct call is.

Perhaps we should reword it to say that the dropDataFrame method gets called in two different ways. One when the data is dropped before adding to the inputQ and the other when the data is dropped through releaseUnconsumed.